### PR TITLE
Selok

### LINKED ISSUE #60: Error messages not implemented in main UI
Errors like when a game was not detected at the target ip should be shown as an in UI error message. Currently they are instead displayed as a popup.

Most critically this allows you to overwrite an existing hero of the same name.

### PR DESCRIPTION
**Pull request by AJenbo**
Fixes #60

Todo: Disallow joining a Nightmare/Hell game with a low-level hero.

This isn't fully pixel-perfect since there where at least 3 different buttons and some odd vertical alignments in tow of the slides for the original, but this is acceptable in this case as it makes them more consistent and was probably just mistaken on the part of the original implementation. Also, this has a very low impact since it's only seen in case of errors. On other difference is that this will slow network as given by the implementation and not try to map them to the original error messages.

### CODE REVIEW

File: SourceX/DiabloUI/selgame.cpp [hunk 1]
```diff
@@ -5,6 +5,7 @@
 #include "DiabloUI/diabloui.h"
 #include "DiabloUI/text.h"
 #include "DiabloUI/dialogs.h"
+#include "selok.h"
```
Comment (glebm) [inline SourceX/DiabloUI/selgame.cpp]: `DiabloUI/selok.h` for consistency

File: SourceX/dvlnet/tcp_server.cpp [hunk 2]
```diff
@@ -82,6 +82,7 @@ void tcp_server::handle_recv(scc con, const asio::error_code &ec,
 				handle_recv_packet(*pkt);
 			}
 		} catch (dvlnet_exception &e) {
+			SDL_Log("Network error");
```
Comment (glebm) [inline SourceX/dvlnet/tcp_server.cpp]: `"Network error: %s", e.what()`

Comment (AJenbo) [reply SourceX/dvlnet/tcp_server.cpp]:
```
/home/ajenbo/code/diablo/devilutionX/SourceX/dvlnet/tcp_server.cpp:85:40: error: ‘virtual const char* dvl::net::dvlnet_exception::what() const’ is private within this context
   85 |    SDL_Log("Network error: %s", e.what());
      |                                        ^

```

Comment (glebm) [reply SourceX/dvlnet/tcp_server.cpp]: `what` is public in the base class so its visiblity should not be narrowed by descendant classes https://en.cppreference.com/w/cpp/error/exception/what

We should fix this in `dvlnet_exception`

File: SourceX/DiabloUI/mainmenu.cpp [hunk 3]
```diff
@@ -1,5 +1,6 @@
 #include "devilution.h"
 #include "DiabloUI/diabloui.h"
+#include "selok.h"
```
Comment (glebm) [inline SourceX/DiabloUI/mainmenu.cpp]: `DiabloUI/selok.h` for consistency here as well

File: SourceX/dvlnet/tcp_server.h [hunk 4]
```diff
@@ -16,6 +16,10 @@ namespace dvl {
 namespace net {
 
 class server_exception : public dvlnet_exception {
+	const char *what() const throw()
```
Comment (glebm) [inline SourceX/dvlnet/tcp_server.h]: Here it should have `public:` and `override`; class methods in C++ are private by default (unlike structs)

File: SourceX/dvlnet/abstract_net.h [hunk 5]
```diff
@@ -13,6 +13,10 @@ namespace net {
 typedef std::vector<unsigned char> buffer_t;
 typedef unsigned long provider_t;
 class dvlnet_exception : public std::exception {
+	const char *what() const throw()
```
Comment (glebm) [inline SourceX/dvlnet/abstract_net.h]:
```c++
 public:
  const char *what() const throw() override
  {
    return "Network error";
  }
```

(`public` and `override`)

File: SourceX/dvlnet/tcp_server.cpp (hunk 2 above)
Comment (AJenbo) [reply SourceX/dvlnet/tcp_server.cpp]: Sorry I have no idea how this work in C++, I tried changing it to `virtual const char *what() const throw()` but it didn't help. I will probably need to expclicity need you to point out the needed change :'D

**Comment by AJenbo** [comment]
@glebm thanks for the review

**Pull request merged**